### PR TITLE
[enh] JBWorker

### PR DESCRIPTION
**Pull request by emdupre**
Updates JBWorker handling, expose commands in RoboNeuro UX.

### CODE REVIEW

**Comment by emdupre** [comment]
Now deploying from this branch with `git push -f heroku enh/jb-worker:master`

Fixes to #8 , #7 , #6  all incorporated. Will confirm on `neurolibre-reviews thread`

**Comment by emdupre** [comment]
Current error :

> 2021-09-27T02:24:12.496657+00:00 app[worker.1]: 4 TID-gogm229t0 WARN: URI::InvalidURIError: bad URI(is not URI?): "\"https://github.com/qmrlab/t1_book\""
2021-09-27T02:24:12.496673+00:00 app[worker.1]: 4 TID-gogm229t0 WARN: /app/vendor/ruby-2.6.3/lib/ruby/2.6.0/uri/rfc3986_parser.rb:67:in `split'
2021-09-27T02:24:12.496673+00:00 app[worker.1]: /app/vendor/ruby-2.6.3/lib/ruby/2.6.0/uri/rfc3986_parser.rb:73:in `parse'
2021-09-27T02:24:12.496673+00:00 app[worker.1]: /app/vendor/ruby-2.6.3/lib/ruby/2.6.0/uri/common.rb:234:in `parse'
2021-09-27T02:24:12.496674+00:00 app[worker.1]: /app/vendor/ruby-2.6.3/lib/ruby/2.6.0/uri/common.rb:737:in `URI'
2021-09-27T02:24:12.496674+00:00 app[worker.1]: /app/neurolibre.rb:53:in `get_repo_name'
2021-09-27T02:24:12.496675+00:00 app[worker.1]: /app/neurolibre.rb:68:in `get_latest_book_build_sha'

**Comment by emdupre** [comment]
calling `roboneuro generate nl-notebook` when none is available successfully fails, but this error isn't propogated to the caller. From the logs :

> 2021-09-27T02:51:01.357205+00:00 app[worker.1]: 4 TID-gr0aawi2k WARN: SystemExit: Missing /binder or /content folder.
2021-09-27T02:51:01.357225+00:00 app[worker.1]: 4 TID-gr0aawi2k WARN: /app/workers.rb:826:in `abort'
2021-09-27T02:51:01.357225+00:00 app[worker.1]: /app/workers.rb:826:in `perform'

But that should be passed back as a GitHub comment in the originating thread.

**Comment by emdupre** [comment]
@agahkarakuzu if I can ask a small q :

In the `NLPreviewWorker` you have `validate_repository_content` hardcoded to check `https://github.com/ltetrel/nha2020-nilearn`. I assumed this was an oversight and am passing the repository address as a param, but this consistently fails `t1_book` with the current validation checks. 

Can you confirm that I should be passing the URL and that the checks should be updated ? Or, is there anything I'm missing here ?

**Comment by agahkarakuzu** [comment]
@emdupre oops, that one was a leftover from debugging, thank you for catching that! 

It should work by passing the repo_address, but it may be expecting the full url, not `repo/user` format. If we are passing that to the `NLPreviewWorker` then `https://github.com/` + `repository_address` may be the solution.

**Comment by emdupre** [comment]
Thanks for confirming !

It's getting the full URL now, but still registering a `Missing /binder or /content folder.`. I was going to try and avoid touching `NLPreviewWorker` or `roboneuro.rb` in this PR, but I might have to :weary: 

Sorry for not catching this in your original PR !

**Comment by emdupre** [comment]
@ltetrel do you have a recommended endpoint to access the listing of all files in the HTML build dir ?

Navigating to `_build` (after logging in) on the server I successfully see the `html` and `jupyter_execute` directories :

Screenshot from 2021-09-28 20-49-27

I can navigate `jupyter_execute` and list the contents :

Screenshot from 2021-09-28 20-49-39

But the the `html` directory returns user-oriented content. I don't think we should remove that view since the preview depends on it, but I need some endpoint that I can direct a GET request towards and iterate over the responses to pull down the built HTML dir. 

Any endpoint suggestions -- or other recommendations -- are appreciated, here !

**Comment by ltetrel** [comment]
Sorry I am not sure if I follow what you mean. I don't control the build layout since this is entirely depending on jupyter book ?

**Comment by ltetrel** [comment]
@emdupre Here you have access to all GET commands available on the API:
http://neurolibre-data.conp.cloud:8081/
If you want to access the notebooks themself, you can download the archive, and search for the notebooks inside the `_build` folder, for example for you:
http://neurolibre-data.conp.cloud/book-artifacts/qmrlab/github.com/t1_book/93466a316bfdb86ae2b0b03afe9d160230b417f1.tar.gz

**Comment by ltetrel** [comment]
Just added a new field `notebook_list` : http://neurolibre-data.conp.cloud:8081/api/v1/resources/books?commit_hash=93466a316bfdb86ae2b0b03afe9d160230b417f1
lmk @emdupre if this is what you wanted.

**Comment by emdupre** [comment]
@ltetrel thanks for your feedback !

To clarify, I don't need access to the notebooks (those are already in the `jupyter_execute` folder, which is accessible). I need access to the generated HTML folder if we're uploading it to GitHub rather than serving from the test server.

That's not available at any of the endpoints you had previously listed. Is it included in the new notebooks endpoint ?

**Comment by ltetrel** [comment]
I don't get what you want ?
Here you have access to all html files: http://neurolibre-data.conp.cloud/book-artifacts/qmrlab/github.com/t1_book/93466a316bfdb86ae2b0b03afe9d160230b417f1/_build/html/
Do you want a list instead is that it ? (like I did for the notebooks)
To download the html, you can download this archive and there will be in the `_build/html` folder: http://neurolibre-data.conp.cloud/book-artifacts/qmrlab/github.com/t1_book/93466a316bfdb86ae2b0b03afe9d160230b417f1.tar.gz

**Comment by emdupre** [comment]
> @emdupre is it ok with this new build_files field? 

Thanks for adding this so quickly, @ltetrel  ! It looks right to me :

Screenshot from 2021-09-30 13-39-52

Based on discussion in #10 , I won't ping this endpoint for now and will just serve the content directly from the test server. But this is great to have for future flexibility !

**Comment by emdupre** [comment]
This branch is currently served through the RoboNeuro app and responsible for all current functionality. I've requested reviews here -- happy for either directly on code or just on function.

When folks are happy we can merge this into main so it's a cleaner deployment !

File: responses/missing_editor_reviewer.erb [hunk 1]
```diff
@@ -1,12 +1,12 @@
-It looks like you don't have an editor and reviewer assigned yet so I can't start the review. Try one or more of these commands:
+It looks like you don't have a manager and tester assigned yet so I can't start the review. Try one or more of these commands:
```
Comment (pbellec) [inline responses/missing_editor_reviewer.erb:1]: review -> test

File: roboneuro.rb [hunk 2]
```diff
@@ -470,12 +472,12 @@ def start_review
     end
 
     if reviewers.empty?
-      respond "Can't start a review without reviewers"
+      respond "Can't start a review without testers"
```
Comment (pbellec) [inline roboneuro.rb:475]: review -> test

**Comment by agahkarakuzu** [comment]
@emdupre are we ready to merge this PR? I will start working on the final components of the workflow, would be cleaner to branch out from main after this is merged.

**Comment by emdupre** [comment]
No we should revert https://github.com/neurolibre/roboneuro/commit/eaafc1b0bfe827ef36a69c301331b45cbc971e69 and subsequent commits to editor / reviewer terminology. I can clean up soon.

**Comment by agahkarakuzu** [comment]
Thank you!

**Comment by emdupre** [comment]
@agahkarakuzu this is now handled in https://github.com/neurolibre/roboneuro/pull/9/commits/60b4db46733bf661cc1c8aa490886f90ffd6bf86 and the PR should be ok to merge. I didn't push to the heroku deployment, though, since I'm unsure where you're currently developing and if you're already pushing there. 

Please feel free to merge in / re-deploy as you're ready !

**Comment by agahkarakuzu** [comment]
Thanks @emdupre! I was working on a fork from this branch, but kept my local changes isolated. I will merge this first, then branch out from `main` to push my changes. 

Thanks!

**Pull request merged**